[17.0] cetmix_tower_server Terminate flight plan

ivs-cetmix opened the pull request:
Terminate a running flight plan. Important: flight plan will be terminated after the current running command is finished.

Forward port of https://github.com/cetmix/cetmix-tower/pull/350

Task: 5011

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Manager-only "Stop" button/action to terminate running flight plans with confirmation and on-screen notification.
  * New "Stopped" status/ribbon and visible indicator for user-terminated plans; stopping a plan also stops running commands.

* **Bug Fixes**
  * Prevents further scheduling/execution after a plan is stopped; cancels queued/running command jobs when stopped.

* **Tests**
  * Added test verifying mid-execution plan stop behavior and resulting statuses.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

GabbasovDinar committed:
[IMP] cetmix_tower_server: Allow stop executing plans

Add new feature:
- Stops the plan instantly

Task: 3410

GabbasovDinar committed:
[IMP] cetmix_tower_server_queue: Allow stop plans

Allow stop executed flight plan in Job queue

Task: 3410

ivs-cetmix commented on cetmix_tower_server_queue/models/cx_tower_server.py:
```diff
@@ -44,3 +44,28 @@ def _command_runner_wrapper(
                 ssh_connection=ssh_connection,
                 **kwargs,
             )
+
+    def _queue_command_runner_wrapper(
+        self,
+        command,
+        log_record,
+        rendered_command_code,
+        sudo=None,
+        rendered_command_path=None,
+        ssh_connection=None,
+        **kwargs,
+    ):
+        # avoid executing command if plan was stopped
+        log_record.invalidate_recordset(["plan_log_id"])
+        if log_record and log_record.plan_log_id and log_record.plan_log_id.is_stopped:
+            return
+
+        return self._command_runner(
+            command=command,
+            log_record=log_record,
+            rendered_command_code=rendered_command_code,
+            sudo=sudo,
+            rendered_command_path=rendered_command_path,
+            ssh_connection=ssh_connection,
+            **kwargs,
+        )
```
@GabbasovDinar is this a relevant suggestion?

GabbasovDinar replied:
@coderabbitai We already reset cash for recordset plan_log_id, so all fields from this record will refreshed

GabbasovDinar replied:
@ivs-cetmix I think we can fix it

ivs-cetmix replied:
Noted, thank you!

ivs-cetmix committed:
[FIX] cetmix_tower_server Stop plans

Ensure that related flight plan is stopped when commmand is stopped.
Remove unused logger import.

ivs-cetmix commented on cetmix_tower_server/models/cetmix_tower.py:
```diff
@@ -223,6 +226,12 @@ def server_check_ssh_connection(
         # Try connecting multiple times
         for attempt in range(1, attempts + 1):
             try:
+                _logger.info(
+                    "Attempt %s of %s to connect to server %s",
+                    attempt,
+                    attempts,
+                    server_reference,
+                )
```
Let's leave is as is for now

ivs-cetmix commented on cetmix_tower_server/ssh/ssh.py:
```diff
@@ -251,20 +251,34 @@ def exec_command(
-
-        exit_status = stdout.channel.recv_exit_status()
-        return exit_status, stdout.readlines(), stderr.readlines()
+        try:
+            stdin, stdout, stderr = ssh_client.exec_command(command)
+            if use_sudo_with_password:
+                stdin.write(self.connection.password + "\n")
+                stdin.flush()
+            exit_status = stdout.channel.recv_exit_status()
+            response = stdout.readlines()
+            error = stderr.readlines()
+            return exit_status, response, error
+        except Exception as e:
+            return 255, [], [str(e)]
 
```
The exception is available in the command log

ivs-cetmix commented on cetmix_tower_server/ssh/ssh.py:
```diff
@@ -356,3 +370,10 @@ def disconnect(self) -> None:
         )
         self.delete_cache(key)
+
+    @classmethod
+    def get_connection_cache(cls):
+        """
+        Get the connection cache.
+        """
+        return cls._connection_cache
```
The access to this data is limited, so it's not leaked

ivs-cetmix commented on cetmix_tower_server/views/cx_tower_plan_log_view.xml:
```diff
@@ -15,13 +26,19 @@
                         name="web_ribbon"
                         title="Finished"
-                        invisible="is_running or not is_running and plan_status != 0"
+                        invisible="is_running or is_stopped or not is_running and plan_status != 0"
+                    />
+                    <widget
```
Let's leave it as is for now

ivs-cetmix commented on cetmix_tower_server/views/cx_tower_plan_log_view.xml:
```diff
@@ -15,13 +26,19 @@
                         title="Failed"
                         bg_color="bg-danger"
-                        invisible="is_running or not is_running and plan_status == 0"
+                        invisible="is_running or is_stopped or not is_running and plan_status == 0"
                     />
```
Let's leave it as is for now

ivs-cetmix committed:
[FIX] cetmix_tower_server_queue Stop commands

Stop command if already stopped flight plan is detected.
Invalidate command log to ensure correct value in flight plan log.

ivs-cetmix commented on cetmix_tower_server_queue/models/cx_tower_server.py:
```diff
@@ -44,3 +44,30 @@ def _command_runner_wrapper(
                 ssh_connection=ssh_connection,
                 **kwargs,
             )
+
+    def _queue_command_runner_wrapper(
+        self,
+        command,
+        log_record,
+        rendered_command_code,
+        sudo=None,
+        rendered_command_path=None,
+        ssh_connection=None,
+        **kwargs,
+    ):
+        # avoid executing command if plan was stopped
+        log_record.invalidate_recordset(["plan_log_id"])
+        if log_record.plan_log_id:
+            log_record.plan_log_id.invalidate_recordset(["is_stopped"])
+        if log_record and log_record.plan_log_id and log_record.plan_log_id.is_stopped:
+            return
+
+        return self._command_runner(
+            command=command,
+            log_record=log_record,
+            rendered_command_code=rendered_command_code,
+            sudo=sudo,
+            rendered_command_path=rendered_command_path,
+            ssh_connection=ssh_connection,
+            **kwargs,
+        )
```
This is a generic wrapper used to implement custom command runners.

jusle reviewed (approved):
Functional LGTM

CetmixGitDrone commented:
This PR has the `approved` label and has been created more than 5 days ago. It should therefore be ready to merge by a maintainer (or a PSC member if the concerned addon has no declared maintainer). 🤖

CetmixGitDrone commented:
This PR has the `approved` label and has been created more than 5 days ago. It should therefore be ready to merge by a maintainer (or a PSC member if the concerned addon has no declared maintainer). 🤖

ivs-cetmix commented on cetmix_tower_server_queue/models/cx_tower_server.py:
```diff
@@ -44,3 +44,30 @@ def _command_runner_wrapper(
                 ssh_connection=ssh_connection,
                 **kwargs,
             )
+
+    def _queue_command_runner_wrapper(
+        self,
+        command,
+        log_record,
+        rendered_command_code,
+        sudo=None,
+        rendered_command_path=None,
+        ssh_connection=None,
+        **kwargs,
+    ):
+        # avoid executing command if plan was stopped
+        log_record.invalidate_recordset(["plan_log_id"])
+        if log_record.plan_log_id:
+            log_record.plan_log_id.invalidate_recordset(["is_stopped"])
+        if log_record and log_record.plan_log_id and log_record.plan_log_id.is_stopped:
+            return
+
+        return self._command_runner(
+            command=command,
+            log_record=log_record,
+            rendered_command_code=rendered_command_code,
+            sudo=sudo,
+            rendered_command_path=rendered_command_path,
+            ssh_connection=ssh_connection,
+            **kwargs,
+        )
```
Noted, thank you!

ivs-cetmix commented:
/ocabot merge patch

CetmixGitDrone commented:
What a great day to merge this nice PR. Let's do it!
Prepared branch [17.0-ocabot-merge-pr-412-by-ivs-cetmix-bump-patch](https://github.com/cetmix/cetmix-tower/commits/17.0-ocabot-merge-pr-412-by-ivs-cetmix-bump-patch), awaiting test results.

CetmixGitDrone commented:
Congratulations, your PR was merged at 2a1ff67fdad15f1306c933af76f9bc57f08d3b8e. Thanks a lot for contributing to cetmix. ❤️
